include/ginkgo/core/matrix/matrix_strategies.hpp

```diff
@@ -0,0 +1,503 @@
+     */
+    virtual void process(const Array<index_type> &mtx_row_ptrs,
+                         Array<index_type> *mtx_srow) = 0;
+
+    /**
+     * Computes the srow size according to the number of nonzeros.
+     *
+     * @param nnz  the number of nonzeros
+     *
+     * @return the size of srow
+     */
+    virtual int64_t calc_size(const int64_t nnz) = 0;
+
+    /**
+     * Copy a strategy. This is a workaround until strategies are revamped,
+     * since strategies like `automatic` do not work when actually shared.
+     */
+    virtual std::shared_ptr<strategy_type> copy() = 0;
+
+protected:
+    void set_name(std::string name) { name_ = name; }
+
+private:
+    std::string name_;
+};
+
+/**
+ * classical is a strategy_type which uses the same number of threads on
+ * each block-row. Classical strategy uses multithreads to calculate on parts of
+ * rows and then do a reduction of these threads results. The number of
+ * threads per row depends on the max number of stored elements per row.
+ */
+template <typename MtxType>
+class classical : public strategy_type<MtxType> {
+public:
+    using index_type = typename strategy_type<MtxType>::index_type;
+
+    /**
+     * Creates a classical strategy.
+     */
+    classical() : strategy_type<MtxType>("classical"), max_length_per_row_(0) {}
+
+    void process(const Array<index_type> &mtx_row_ptrs,
+                 Array<index_type> *mtx_srow) override
+    {
+        auto host_mtx_exec = mtx_row_ptrs.get_executor()->get_master();
+        Array<index_type> row_ptrs_host(host_mtx_exec);
+        const bool is_mtx_on_host{host_mtx_exec == mtx_row_ptrs.get_executor()};
+        const index_type *row_ptrs{};
+        if (is_mtx_on_host) {
+            row_ptrs = mtx_row_ptrs.get_const_data();
+        } else {
+            row_ptrs_host = mtx_row_ptrs;
+            row_ptrs = row_ptrs_host.get_const_data();
+        }
+        auto num_rows = mtx_row_ptrs.get_num_elems() - 1;
+        max_length_per_row_ = 0;
+        for (index_type i = 1; i < num_rows + 1; i++) {
+            max_length_per_row_ =
+                std::max(max_length_per_row_, row_ptrs[i] - row_ptrs[i - 1]);
+        }
+    }
+
+    int64_t calc_size(const int64_t nnz) override { return 0; }
+
+    index_type get_max_length_per_row() const noexcept
+    {
+        return max_length_per_row_;
+    }
+
+    std::shared_ptr<strategy_type<MtxType>> copy() override
+    {
+        return std::make_shared<classical<MtxType>>();
+    }
+
+private:
+    index_type max_length_per_row_;
+};
+
+/**
+ * load_balance is a strategy_type which uses the load balance algorithm.
+ */
+template <typename MtxType>
+class load_balance : public strategy_type<MtxType> {
+public:
+    using index_type = typename strategy_type<MtxType>::index_type;
+
+    /**
+     * Creates a load_balance strategy.
+     */
+    load_balance()
+        : load_balance(std::move(
+              gko::CudaExecutor::create(0, gko::OmpExecutor::create())))
```

Comment:
This is actually exactly the same as what is done for CSR. The script copied all this to fbcsr.hpp, but it was decided not use the strategies for fbcsr. I didn't want to get rid of this yet, so I just moved it to a different file but templated it on the matrix type. My thinking was that at some point I would have specializations for Fbcsr matrices and Csr could use this file too. I'm not sure how much commonality would be there between the two but there might be some.

The other options is I rename this file and namespace to `fbcsr_strategies`, remove the templating on the matrix type, and only Fbcsr uses it. Maybe this is better.

But it's very tentative; if you think this is problematic, it could be completely removed for now.